### PR TITLE
Add a comment about locking a `Mutex` multiple times

### LINKED ISSUE #32260: document reentrancy of `std::sync::Mutex`.
`std::sync::Mutex` is not reentrant (you can't `.lock` a mutex if you already `.lock`ed it and haven't released the lock), but this behavior is [not documented](https://doc.rust-lang.org/nightly/std/sync/struct.Mutex.html). 

Looks like the exact consequences of locking the mutex twice vary by platform. On linux, you get a deadlock. On windows there is a panic: https://github.com/rust-lang/rust/blob/master/src/libstd/sys/windows/mutex.rs#L75. Maybe this should `panic`  on all platforms?

### PR DESCRIPTION
**Pull request by tbu-**
Fixes #32260.

### CODE REVIEW

**Comment by alexcrichton** [comment]
A guarantee that we *can* provide here is that a recursive call to `lock` will never return (as that wouldn't be sound otherwise). We could then perhaps just note that this is implemented via deadlocking on some platforms and panicking on others?

**Comment by tbu-** [comment]
@alexcrichton Updated.

File: src/libstd/sync/mutex.rs [hunk 1]
```diff
@@ -205,10 +205,19 @@ impl<T: ?Sized> Mutex<T> {
     /// held. An RAII guard is returned to allow scoped unlock of the lock. When
     /// the guard goes out of scope, the mutex will be unlocked.
     ///
+    /// The exact behavior on locking a mutex twice in the same thread is left
```
Comment (nagisa) [inline src/libstd/sync/mutex.rs]: “locking a mutex twice” is ambiguous in whether 2nd lock happens with mutex being locked or unlocked.

I would reword this as “behaviour of locking the mutex in the thread which already holds the lock is left unspecified” instead.

**Comment by tbu-** [comment]
@nagisa Sounds good. Changed.

**Comment by nagisa** [comment]
@bors r+ rollup df748bf

Seems good
